protocols/relay/src/v2/client.rs

```diff
@@ -0,0 +1,561 @@
+                                .push(RqstPendingConnection::Reservation { to_listener });
+                            let handler = self.new_handler();
+                            return Poll::Ready(NetworkBehaviourAction::Dial {
+                                opts: DialOpts::peer_id(relay_peer_id)
+                                    .addresses(vec![relay_addr])
+                                    .extend_addresses_through_behaviour()
+                                    .build(),
+                                handler,
+                            });
+                        }
+                    }
+                }
+                Poll::Ready(Some(transport::TransportToBehaviourMsg::DialReq {
+                    relay_addr,
+                    relay_peer_id,
+                    dst_peer_id,
+                    send_back,
+                    ..
+                })) => {
+                    match self
+                        .connected_peers
+                        .get(&relay_peer_id)
+                        .and_then(|cs| cs.get(0))
+                    {
+                        Some(connection_id) => {
+                            return Poll::Ready(NetworkBehaviourAction::NotifyHandler {
+                                peer_id: relay_peer_id,
+                                handler: NotifyHandler::One(*connection_id),
+                                event: handler::In::EstablishCircuit {
+                                    send_back,
+                                    dst_peer_id,
+                                },
+                            });
+                        }
+                        None => {
+                            self.rqsts_pending_connection
+                                .entry(relay_peer_id)
+                                .or_default()
+                                .push(RqstPendingConnection::Circuit {
+                                    dst_peer_id,
+                                    send_back,
+                                });
+                            let handler = self.new_handler();
+                            return Poll::Ready(NetworkBehaviourAction::Dial {
+                                opts: DialOpts::peer_id(relay_peer_id)
+                                    .addresses(vec![relay_addr])
+                                    .extend_addresses_through_behaviour()
+                                    .build(),
+                                handler,
+                            });
+                        }
+                    }
+                }
+                Poll::Ready(None) => unreachable!(
+                    "`Relay` `NetworkBehaviour` polled after channel from \
+                     `RelayTransport` has been closed.",
+                ),
+                Poll::Pending => break,
+            }
+        }
+
+        Poll::Pending
+    }
+}
+
+/// A [`NegotiatedSubstream`] acting as a [`RelayedConnection`].
+pub enum RelayedConnection {
```

Comment:
May I suggest this refactor? https://github.com/MarcoPolo/rust-libp2p/commit/039803134e484a74ce0c528bb7593a5522762327

It removes the Poisoned state and simplifies the logic a bit.